mitiq/shadows/quantum_processing.py

```diff
@@ -0,0 +1,140 @@
+    qubits = sorted(list(circuit.all_qubits()))
+    rotated_circuits = []
+    for pauli_string in pauli_strings:
+        rotated_circuit = circuit.copy()
+        for qubit, pauli in zip(qubits, pauli_string):
+            # Pauli X measurement is equivalent to H plus a Z measurement
+            if pauli == "X":
+                rotated_circuit.append(cirq.H(qubit))
+            # Pauli X measurement is equivalent to S^-1*H plus a Z measurement
+            elif pauli == "Y":
+                rotated_circuit.append(cirq.S(qubit) ** -1)
+                rotated_circuit.append(cirq.H(qubit))
+            # Pauli Z measurement
+            else:
+                assert (
+                    pauli == "Z"
+                ), f"Pauli must be X, Y, Z. Got {pauli} instead."
+        if add_measurements:
+            rotated_circuit.append(cirq.measure(*qubits))
+        rotated_circuits.append(rotated_circuit)
+    return rotated_circuits
+
+
+def random_pauli_measurement(
+    circuit: cirq.Circuit,
+    n_total_measurements: int,
+    executor: Callable[[cirq.Circuit], MeasurementResult],
+) -> Tuple[NDArray[Any], NDArray[Any]]:
+    r"""
+    Given a circuit, perform random Pauli measurements on the circuit and
+    return outcomes. The outcomes are represented as a string where a
+    z-basis measurement outcome of :math:`|0\rangle` corresponds to 1, and
+    :math:`|1\rangle` corresponds to -1.
+
+    Args:
+        circuit: Cirq circuit.
+        n_total_measurements: number of snapshots.
+        executor: A callable which runs a circuit and returns a single
+            bitstring. If more bitstrings are returned, only one of them
+            is actually used.
+
+    Returns:
+        Tuple of two numpy arrays. The first array contains
+        measurement outcomes (-1, 1) while the second array contains the
+        indices for the sampled Pauli's ("X", "Y", "Z").
+        This implies that local Clifford rotations plus z-basis measurements
+        are effectively equivalent to random Pauli measurements.
+        Each row of the arrays corresponds to a distinct snapshot or sample,
+        while each column corresponds to measurement outcomes
+        and random Pauli measurement on a different qubit.
+    """
+
+    num_qubits = len(circuit.all_qubits())
+    pauli_strings = generate_random_pauli_strings(
+        num_qubits, n_total_measurements
+    )
+
+    # Rotate and attach measurement gates to the circuit
+    rotated_circuits = get_rotated_circuits(
+        circuit,
+        pauli_strings,
+    )
+    if tqdm is not None:
+        rotated_circuits = tqdm(
+            rotated_circuits,
+            desc="Measurement",
+            leave=False,
+        )
+    results = [
+        executor(rotated_circuit) for rotated_circuit in rotated_circuits
+    ]
+
+    # Transform the outcomes into a numpy array 0 -> 1, 1 -> -1.
+    shadow_outcomes = []
+    for result in results:
+        bitstring = list(result.get_counts().keys())[0]
```

Comment:
Should we throw an error here if `len(result.get_counts().keys()) > 1`?

Reply:
Makes sense to me. It is quite easy to forget about the single-shot assumption. 

Note that, if we throw an error, the previously suggested warning should be changed into something like this:
```
Warning:
    The ``executor`` must return a ``MeasuremenrResult``
    for a single shot (i.e. a single bitstring).
```